### PR TITLE
add std::net::url for parsing/generating URLs

### PR DESCRIPTION
**Pull request by louis77**
Tests are inferred from https://datatracker.ietf.org/doc/html/rfc3986

### CODE REVIEW

File: lib/std/core/string.c3 [hunk 1]
```diff
@@ -362,6 +362,29 @@ fn usz! String.index_of_char(s, char needle)
 	return SearchResult.MISSING?;
 }
 
+<*
+ Find the index of the first incidence of a one of the chars.
+
+ @param [in] s
+ @param needle "The characters to look for"
```
Comment (lerno) [inline lib/std/core/string.c3:369]: needle can also be [in] since it is not mutated.

File: lib/std/net/url.c3 [hunk 2]
```diff
@@ -0,0 +1,189 @@
+module std::net::url;
+
+import std::io;
+import std::collections::map;
+import std::collections::list;
+
+def ListStr = List(<String>);
+def Values = map::HashMap(<String, ListStr>);
+
+struct Url(Printable) {
```
Comment (lerno) [inline lib/std/net/url.c3:10]: The stdlib uses Allman brace style, so please change it to make it consistent.

File: lib/std/net/url.c3 [hunk 3]
```diff
@@ -0,0 +1,189 @@
+module std::net::url;
+
+import std::io;
+import std::collections::map;
+import std::collections::list;
+
+def ListStr = List(<String>);
+def Values = map::HashMap(<String, ListStr>);
```
Comment (lerno) [inline lib/std/net/url.c3:8]: `map::HashMap` can be replaced by simply `HashMap`

File: lib/std/net/url.c3 [hunk 4]
```diff
@@ -0,0 +1,189 @@
+module std::net::url;
+
+import std::io;
+import std::collections::map;
+import std::collections::list;
+
+def ListStr = List(<String>);
```
Comment (lerno) [inline lib/std/net/url.c3:7]: It is more common in the stdlib to swap these, so `StrList` or `StringList`

File: lib/std/net/url.c3 [hunk 5]
```diff
@@ -0,0 +1,189 @@
+module std::net::url;
+
+import std::io;
+import std::collections::map;
+import std::collections::list;
```
Comment (lerno) [inline lib/std/net/url.c3:5]: These three can be collected in a single import:
`import std::io, std::collections::map, std::collections::list;`

File: lib/std/net/url.c3 [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    String password;
+    String path;
+    String query;
+    String fragment;
+}
+
+fn Url! parse(String url_string) {
+    Url url = Url{};
+    url_string = url_string.trim();
+    if (url_string.len == 0) {
```
Comment (lerno) [inline lib/std/net/url.c3:24]: More idiomatic `if (!url_string) return url;`

File: lib/std/net/url.c3 [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+        return url;
+    }
+
+    // Parse scheme
+    if (try pos = url_string.index_of("://")) {
+        url.scheme = url_string[0 .. pos-1];
```
Comment (lerno) [inline lib/std/net/url.c3:30]: More idiomatic: `url.scheme = url_string[:pos];`

File: lib/std/net/url.c3 [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        url_string = url_string[url.scheme.len + 3..];
+    }
+    else if (url_string.contains(":")) {
+        // Handle schemes without authority like 'mailto:'
+        url.scheme = url_string[0 .. url_string.index_of(":")!-1];
```
Comment (lerno) [inline lib/std/net/url.c3:35]: Similar here (and later): if you find yourself writing `a[0..y - 1]` you can instead write this as `a[:y]`

File: lib/std/net/url.c3 [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        url_string = url_string[url.scheme.len+1..];
```
Comment (lerno) [inline lib/std/net/url.c3:36]: Spacing around the operators to match stdlib code style.

File: lib/std/net/url.c3 [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+            defer free(userpass);
+            url.username = userpass[0];
+            if (userpass.len > 1) {
+                url.password = userpass[1];
+            }
+            authority = authority[userInfoEnd+1..];
+        }
+
+        // Check for IPv6 address in square brackets
+        if (authority.starts_with("[") && authority.contains("]")) {
```
Comment (lerno) [inline lib/std/net/url.c3:63]: Consider replacing if / else with a switch

File: lib/std/net/url.c3 [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    return url;
+}
+
+fn String Url.to_string(Url* self, Allocator allocator = allocator::temp()) @dynamic
+{
+    @pool(allocator)
+	{
+        DString builder = dstring::new("", allocator);
```
Comment (lerno) [inline lib/std/net/url.c3:120]: Use the temp allocator for the dstring instead?

File: lib/std/net/url.c3 [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+            builder.append_chars(self.query);
+        }
+
+        // Add fragment if it exists
+        if (self.fragment != "") {
+            builder.append_char('#');
+            builder.append_chars(self.fragment);
+        }
+
+        return builder.str_view();
```
Comment (lerno) [inline lib/std/net/url.c3:164]: This is returning a str_view from a freed builder and is a use after free.
Instead, make the dstring a temp dstring using temp_init. Then 
`return builder.copy_str(allocator);`

File: lib/std/net/url.c3 [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+    };
+}
+
+fn Values Url.query_values(Url* self, Allocator allocator = allocator::heap()) {
+    Values vals = {};
+
+    String[] raw_vals = self.query.split("&");
```
Comment (lerno) [inline lib/std/net/url.c3:171]: Consider using Splitter using `self.query.splitter("&")` instead. You might choose .tokenizer if you want to skip empty parts (eg `&&`) from being picked up.

File: lib/std/net/url.c3 [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+            builder.append_char('#');
+            builder.append_chars(self.fragment);
+        }
+
+        return builder.str_view();
+    };
+}
+
+fn Values Url.query_values(Url* self, Allocator allocator = allocator::heap()) {
+    Values vals = {};
```
Comment (lerno) [inline lib/std/net/url.c3:169]: You need to initialize the hash map here with the allocator, so use:
```c
Values vals;
vals.init(allocator);
```

File: lib/std/core/string.c3 (hunk 1 above)
Comment (louis77) [reply lib/std/core/string.c3]: Done.

File: lib/std/net/url.c3 (hunk 2 above)
Comment (louis77) [reply lib/std/net/url.c3]: Done.

File: lib/std/net/url.c3 (hunk 3 above)
Comment (louis77) [reply lib/std/net/url.c3]: Done.

File: lib/std/net/url.c3 (hunk 4 above)
Comment (louis77) [reply lib/std/net/url.c3]: Done.

File: lib/std/net/url.c3 (hunk 5 above)
Comment (louis77) [reply lib/std/net/url.c3]: Done.

File: lib/std/net/url.c3 (hunk 6 above)
Comment (louis77) [reply lib/std/net/url.c3]: Done.

File: lib/std/net/url.c3 (hunk 7 above)
Comment (louis77) [reply lib/std/net/url.c3]: Done.

File: lib/std/net/url.c3 (hunk 8 above)
Comment (louis77) [reply lib/std/net/url.c3]: Done.

File: lib/std/net/url.c3 (hunk 9 above)
Comment (louis77) [reply lib/std/net/url.c3]: Done.

File: lib/std/net/url.c3 (hunk 12 above)
Comment (louis77) [reply lib/std/net/url.c3]: Done.

File: lib/std/net/url.c3 (hunk 13 above)
Comment (louis77) [reply lib/std/net/url.c3]: Done, I used the tokenizer.

File: lib/std/net/url.c3 (hunk 14 above)
Comment (louis77) [reply lib/std/net/url.c3]: Done.

File: lib/std/net/url.c3 (hunk 10 above)
Comment (louis77) [reply lib/std/net/url.c3]: As discussed, for future consideration.

**Pull request closed without merging**